### ISSUE TITLE
Add Umbraco support for TinyMCE wrapper styles

### ISSUE THREAD

**Issue by c9mbundy** [umbraco/Umbraco-CMS#5539]
There was a previous issue dealing with this: U4-6267 - Support advanced rule set styles for tinymce rte editor

However, it appears to have fallen by the wayside, and I can't see that there is a standard way to implement the wrapper styles that TinyMCE supports.

e.g. I have a client who wants to be able to mark a selection of paragraphs and then apply a 'style' to them which will wrap them in a div that applies a class (or classes) that implements multiple-column layout for them using the standard W3C CSS for this - i.e. select a 'Columns-2' style.

Of course, this is just one of many obvious requirements to be able to apply a Wrapper style to a selection of elements - albeit more commonly a selection of paragraphs.

At the moment there appears to only be (documented) support in the selector syntax used for the style dropdown to apply TinyMCE's InLine and Block styles, but neither of these will give you the ability to wrap a selection of elements inside a single element, which the TinyMCE Wrapper style does.

**Comment by c9mbundy** [comment]
Just following this up with a further comment.

Previous issue U4-6267 was focused on devising a custom syntax for advanced rules to control how the members of the the style_format dropdown in TinyMCE implement the advanced options detailed here:
[https://www.tiny.cloud/docs/configure/content-formatting/](url)
[https://www.tiny.cloud/docs/configure/editor-appearance/#style_formats](url)

However, I would contend that trying to manipulate the css syntax to achieve this might not be the best approach, as it would be non-standard and non-intuitive.

Umbraco already uses a reserved comment for controlling one of the style_format attribute parameters (title) so why not simply extend that approach to deal with more style_format features available in TinyMCE?

e.g. umb_wrapper:<true|false> - Use the style_format as a Wrapper for the selection
       umb_exact:<true|false> - Do not merge with other wrappers having the same format
       umb_group:<string> - Locate this style_format below a sub-group heading (create if required)
       umb_nested:<string> - Locate this style_format in a sub-menu (create if required) 
       
Basically, any of the formats, attributes and parameters supported by TinyMCE and seen as desirable to Umbraco admins could be specified as reserved comments (separated by ; semi-colons) that are already being parsed to provide the Title in the drop-down.

**Comment by nul800sebastiaan** [comment]
Hi @c9mbundy - I recognize your need for specialized markup for the content that you have. What you're basically proposing is to have a complete site build in the rich text editor. You've noticed that there's a lot of friction in accomplishing this as well, as you've seen from the TinyMCE docs, there's no real good ways to do this.

I would encourage you to look into the Grid and setting it up so that your editors can actually set up rows and columns on their own, each grid cell can contains a rich text editor to suit all your needs.

We currently have no plans to enhance the rich text editor for building layouts like you propose.

**Comment by c9mbundy** [comment]
Hi @nul800sebastiaan  - based on your comment, I can only assume that you've misread the issue and have thus missed the point.

This is **NOT** about adding layout functionality to the editor, but is about adding support for features in the 'formats' drop-down of TinyMCE, that are currently not supported by Umbraco, but which are supported by others such as WordPress.

TinyMCE supports a range of standard parameters and attributes for managing style entries in the formats dropdown, and for controlling how those styles should be applied.

Among those is the ability to indicate that a style should be applied as a Wrapper element.class to a selection, rather than modify multiple block elements (as Umbraco currently does) and this give the content admin/editor more control over their use of content styling.

Umbraco simply requires a way to indicate this requirement should be used by the style, as the stylesheet for the formats dropdown is applied to the editor, so that TinyMCE knows how the style should be applied.

Umbraco is already using a solution to manage one of TinyMCE's parameter requirements - that items in the formats dropdown must have a title - by using a reserved comment of '/**umb_name:' in the stylesheet applied to the editor.

I'm simply suggesting that other parameters supported by TinyMCE's style_formats (the formats dropdown) - such as the ability to group and/or nest styles in the dropdown - could also be indicated and parsed by extending this existing approach with other comment-based commands. e.g. umb_group, umb_nested, umb_wrapper, etc etc.

This is **NOT** about html layout, but is about presentation, and is about adding styles to the formats dropdown of TinyMCE that content admins/editors currently don't have access to, because Umbraco does not fully support the styling features of TinyMCE, in the way that say, WordPress does.

The example given just happened to be a request I have for W3C multi-column CSS (and for the record, grids are not a replacement for auto-flow multi-column CSS) but it could just as easily have been any css that needs to be applied by a content admin/editor to a selection, by using a wrapper element.class - e.g. a blockquote.class that needs to wrap a selection paragraphs - rather than the current limitation of selections being modified as multiple block elements, or applying multiple spans, that is often either ineffective or quite simply wrong, and requiring content admins/editors to manually edit html to correct - something that is often beyond their ability.

The previous issue U4-6267 attempted to address this requirement by introducing a non-standard css syntax, but I am suggesting it could be addressed by extending the existing use of reserved comments, allowing management of the styles in the formats dropdown from the stylesheet used for this purpose, rather than needing a developer to create special code that is difficult to manage.

**Comment by nielslyngsoe** [comment]
Hi @c9mbundy 

There is a way in v7 which is still supported in v8.

What you can do is to setup these more advanced styles in the config/TinyMCEConfig.config

In the bottom of the file you have the <customConfig> definition, where you can add the style_formats configuration and specify styles a bit more in details. See this example:

```
  <customConfig>
    <config key="style_formats">
        [{"title": "Two column layout", "block":"div", "classes":"columns-2", 
        "wrapper": true}]
    </config>
  </customConfig>
```

In this way you'r able to make styles that wraps content — You can read more about the options here:
https://www.tiny.cloud/docs/configure/content-formatting/

I hope that can lead you to a solution.

**Comment by c9mbundy** [comment]
Hi @nielslyngsoe 

Thanks for that input, and yes you are quite correct - it can be done that way, and I have resorted to that method previously.

However, I'm not aware that there is a way to specify which data-type should use such a custom config, and so unless I'm mistaken, the trade-off there is that such a config defines the style_formats for every instance of TinyMCE used by Umbraco, whereas loading a stylesheet that can be parsed and used to provide that configuration detail, means that it can be varied between different instances and applications of the editor. i.e. A "one-size-fits-all" solution may or may-not be acceptable to a client needing strict control over editing formats, while a stylesheet based solution would allow you to include styles that make sense for some applications of the editor, and not for others.

**Comment by nielslyngsoe** [comment]
@c9mbundy Good point, makes your case stronger.

I will ask @nul800sebastiaan to reconsider our options for this feature, I do not think we have the resources to implement this feature within reasonable time. But if anyone has the ability to make a PR for it we will accept it.

**Comment by nul800sebastiaan** [comment]
So I had a bit of a chat with @nielslyngsoe about this trying to understand the use case for this, we found out that all he could think of is possible already. Consider the following RTE content, all of these paragraphs used to be regular `<p>` tags, nothing special about them. With the added RTE stylesheet an applying different pieces of markup to it, it looks like this now: 

![image](https://user-images.githubusercontent.com/304656/58864658-6c8ea100-86b5-11e9-963e-652a2dc47a39.png)

The HTML output of that looks like this:

```
<h1>Lorem ipsum dolor sit amet, consectetur adipiscing elit. <span class="marked">Nullam eget lacinia nisl.</span> Aenean sollicitudin diam vitae enim ultrices, semper euismod magna efficitur.</h1>
<p>Here is where the quote goes:</p>
<div class="quote">"TinyMCE <span class="marked">does so</span> much"</div>
<p> </p>
<p>Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Integer sit amet feugiat sapien, sed dictum ante. <span class="marked">Praesent venenatis <span class="farked">tristique</span> pellentesque</span>. Vestibulum tincidunt magna vel ante tristique, et cursus <a href="/{localLink:umb://document/ca4249ed2b234337b52263cabe5587d1}" title="Home">neque <span class="marked">eleifend</span></a><span class="marked">. Nulla elit a</span>nte, aliquam eu pulvinar sit amet, accumsan eu eros. Mauris felis odio, congue et ipsum vitae, ultricies dapibus risus. In non viverra ex. Nam lorem justo, auctor eget imperdiet a, dignissim nec mauris.</p>
```

And for reference, my RTE stylesheet looks like:

```
/**umb_name:h1 red*/
h1 {
	color:red;
}

/**umb_name:marked*/
.marked {
	background-color: yellow;
}

/**umb_name:farked*/
.farked {
	text-decoration: line-through;
}

/**umb_name:quote*/
div.quote {
	color: green;
}
```

And yes, this doesn't make it possible to take multiple paragraphs and wrap a div around it but I think Niels and I agree that if you need specialized markup like that then you should look into using the Grid or alternatives for that. 

FYI, we know the Grid isn't perfect but it'll work for now. We're looking to create new editing experiences in the future that will be even better, see: https://github.com/umbraco/rfcs/pull/11

So with specialized text markup being very well possible and with block level support in the Grid, we will close this once again, sorry to disappoint but hopefully you've got some inspiration now to improve the editor experience!

**Comment by nielslyngsoe** [comment]
I think we should underline that by defining "div.MyClass" in your stylesheet, it will work as a wrapper that can wrap other styles, in this example the ".marked". Which I guess would be solving your case?
